### PR TITLE
HLL coverage improvements

### PR DESCRIPTION
**Pull request by luizirber**
## HLL coverage improvements

HyperLogLog line coverage is a bit low because some conditions during estimation are not so easy to trigger. After exposing the HLL internals it was easier to find test cases using data already available for the current tests.

This PR improves the coverage and also simplifies some error checking inside the HLL.

### CODE REVIEW

**Comment by luizirber** [comment]
- [ ] Is it mergable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage.
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Is it documented in the ChangeLog?
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

**Comment by luizirber** [comment]
Ready for review @mr-c @ctb @camillescott

File: khmer/_khmermodule.cc [hunk 1]
```diff
@@ -4337,6 +4330,41 @@ static PyObject * hllcounter_consume_fasta(khmer_KHLLCounter_Object * me,
...
+
+static
+PyObject *
+hllcounter_getcounters(khmer_KHLLCounter_Object * me)
+{
+    std::vector<int> counters = me->hllcounter->get_M();
+
+    PyObject * x = PyList_New(counters.size());
+    for (size_t i = 0; i < counters.size(); i++) {
+        PyList_SET_ITEM(x, i, PyLong_FromLong(counters[i]));
```
Comment (luizirber) [inline khmer/_khmermodule.cc]: I was really tempted to use the buffer interface, but I think it is better to do it when implementing #714

File: lib/hllcounter.cc [hunk 2]
```diff
@@ -32,20 +32,14 @@ std::map<int, std::vector<double> > rawEstimateData;
...
-        std::stringstream message;
-        if (p < 4) {
-            message << "Max error should be smaller than " << valid_upper_bound;
-        } else {
-            message << "Min error should be greater than " << valid_lower_bound;
-        }
-        throw khmer_exception(message.str().c_str());
+    if (p < 4) {
+        // ceil(log2((1.04 / x) ^ 2)) = 4, solve for x
+        throw khmer_exception("Error rate must be smaller than 0.367696");
```
Comment (mr-c) [inline lib/hllcounter.cc]: Is this requirement always or only under certain conditions?

Comment (luizirber) [inline lib/hllcounter.cc]: Always. Technically you can make the HLL have a even smaller memory footprint using p < 4, but both the original paper and the Google extension of the original idea limit it to p > 4. If the error is bigger than 0.367696 the equation in the previous line will result in p < 4.

Comment (mr-c) [inline lib/hllcounter.cc]: The error message leaves a naive reader confused: what did they do wrong and how can they fix it?

Comment (luizirber) [inline lib/hllcounter.cc]: This is how it shows up in Python:

```python
In [1]: import khmer
In [2]: khmer.HLLCounter(0.5, 32)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-d2705791ce3b> in <module>()
----> 1 khmer.HLLCounter(0.5, 32)

ValueError: Error rate must be smaller than 0.367696

In [3]: khmer.HLLCounter(0.000000001, 32)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-67d189173ec1> in <module>()
----> 1 khmer.HLLCounter(0.000000001, 32)

ValueError: Error rate must be greater than 0.0040624
```
Two options:
- Change the error message to be something like "Please set error rate to a value smaller than 0.367696"
- Force it to be at the limit (either upper or lower) and only emit a warning.

File: ChangeLog [hunk 3]
```diff
@@ -1,3 +1,9 @@
+2015-01-25  Luiz Irber  <irberlui@msu.edu>
+    * khmer/_khmermodule.cc: expose HLL internals as read-only attributes.
```
Comment (mr-c) [inline ChangeLog]: newline please

File: lib/hllcounter.cc (hunk 2 above)
Comment (mr-c) [inline lib/hllcounter.cc]: To recap our discussion: My recommendation is:
- [ ] add default values for error rate and k-size to the tp_new function
- [ ] add setters for these with docstrings that only work prior to first counting
- [ ] document the above in a Python visible way; check `import khmer; doc(khmer.HLLCounter)` for the user view.

File: khmer/_khmermodule.cc (hunk 1 above)
Comment (maarten1983) [inline khmer/_khmermodule.cc]: pyobject+

**Comment by mr-c** [comment]
@luizirber As soon as this is made mergable and https://github.com/ged-lab/khmer/pull/738#discussion_r23560245 is address I think it will be ready to go

**Comment by luizirber** [comment]
Jenkins, test this please

**Comment by luizirber** [comment]
- [ ] Is it mergeable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage with `make clean diff-cover`
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Is it documented in the ChangeLog?
  http://en.wikipedia.org/wiki/Changelog#Format
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

**Comment by luizirber** [comment]
Jenkins, test this please

**Comment by luizirber** [comment]
Ready for review @mr-c @ctb @camillescott

File: khmer/_khmermodule.cc [hunk 4]
```diff
@@ -4480,6 +4578,38 @@ static PyMethodDef khmer_hllcounter_methods[] = {
...
+        NULL
+    },
+    {
+        (char const *)"counters",
+        (getter)hllcounter_getcounters, NULL,
+        (char const *)"Read-only internal counters.",
+        NULL
+    },
+    {NULL} /* Sentinel */
+};
```
Comment (mr-c) [inline khmer/_khmermodule.cc]: `khmer/_khmermodule.cc:4611:1: warning: deprecated conversion from string constant to ‘char*’ [-Wwrite-strings]` ?

File: khmer/_khmermodule.cc [hunk 5]
```diff
@@ -4455,6 +4448,111 @@ static PyObject * hllcounter_consume_fasta(khmer_KHLLCounter_Object * me,
...
+                     void *closure)
+{
+    if (value == NULL) {
+        PyErr_SetString(PyExc_TypeError, "Cannot delete attribute");
+        return -1;
+    }
+
+    long ksize = 0;
+    if (PyLong_Check(value)) {
+        ksize = PyLong_AsLong(value);
```
Comment (mr-c) [inline khmer/_khmermodule.cc]: No code coverage for line 4477

File: khmer/_khmermodule.cc [hunk 6]
```diff
@@ -4455,6 +4448,111 @@ static PyObject * hllcounter_consume_fasta(khmer_KHLLCounter_Object * me,
...
+        PyErr_SetString(PyExc_ValueError, "Please set k-mer size to a value "
+                        "greater than zero");
+        return -1;
+    }
+
+    try {
+        me->hllcounter->set_ksize(ksize);
+    } catch (InvalidValue &e) {
+        PyErr_SetString(PyExc_ValueError, e.what());
+        return -1;
```
Comment (mr-c) [inline khmer/_khmermodule.cc]: No code coverage for lines 4494-4495

**Comment by mr-c** [comment]
Can you eliminate that warning and cover those last two bits with tests?

Otherwise, it LGTM; I love the output of `python -c 'import khmer; help(khmer.HLLCounter)'`!

File: khmer/_khmermodule.cc (hunk 4 above)
Comment (luizirber) [inline khmer/_khmermodule.cc]: So, I don't really get **where** is it coming from. Is it from the sentinel?

Comment (mr-c) [inline khmer/_khmermodule.cc]: I get multiple warnings for the same line number so I'm guessing it is all the conversions within the code block.

Comment (mr-c) [inline khmer/_khmermodule.cc]: I get multiple warnings for the same line number so I'm guessing it is all
the conversions within the code block.

Comment (luizirber) [inline khmer/_khmermodule.cc]: Found it, shouldn't have used const.

**Comment by luizirber** [comment]
jenkins, Test this please

**Comment by luizirber** [comment]
Jenkins, test this please

**Comment by luizirber** [comment]
@mr-c fixed remaining issues

**Pull request merged**